### PR TITLE
feat: Prims algorithm

### PR DESCRIPTION
**Pull request by Sl1mb0**
#### Description of Change

I implemented prim's algorithm and added `prim.c` to `TheAlgorithms/C/greedy_approach`. 

Prim's algorithm finds a minimum spanning tree of a weighted, connected graph. 
Using a 'greedy' approach to generate the MST, the algorithm searches for a minimally weighted, non-visited edge, and adds the vertex connected by that edge to the MST.

To test, compile and run.
Enter in adjacency matrix of a weighted, connected graph.

#### References

pg. 319 "Introduction to the Design and Analysis of Algorithms" - Anany Levitin

#### Checklist

- [x ] Added description of change
- [x ] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [ x] Added tests and example, test must pass
- [ x] Relevant documentation/comments is changed or added
- [ x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [ x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Comment by Sl1mb0** [comment]
Who needs to review the code I've added? What exactly are they reviewing for? My code has passed the automated tests.

**Review by ayaankhan98 — COMMENTED** [review]
Hey @Sl1mb0, Thanks for your contribution.
Please follow the [Repository Standard Guidelies](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md) strictly.

**Comment by Sl1mb0** [comment]
Did I not follow the guidelines? I can't find an issue.

**Comment by KirwinD** [comment]
@Sl1mb0  great input.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
There is no proper [documentation](https://www.doxygen.nl/manual/docblocks.html), and the code is not up to the [repository standards](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md). Please follow the steps documented. You can see #797 as a reference. Thanks. 🙂

File: greedy_approach/prim.c [hunk 1]
```diff
@@ -0,0 +1,207 @@
+/**
+ * @file
+ * @author [Timothy Maloney] (https://github.com/sl1mb0)
```
Comment (Panquesito7) [inline greedy_approach/prim.c:3]:
```suggestion
 * @author [Timothy Maloney](https://github.com/sl1mb0)
```

File: greedy_approach/prim.c [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+ * @brief [Prim's algorithm] (https://en.wikipedia.org/wiki/Prim%27s_algorithm)
```
Comment (Panquesito7) [inline greedy_approach/prim.c:4]:
```suggestion
 * @brief [Prim's algorithm](https://en.wikipedia.org/wiki/Prim%27s_algorithm) implementation
```

File: greedy_approach/prim.c [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+
+#define MAX 20
+#define INF 999
+
+int G[MAX][MAX];    // weighted, connected graph G
+int MST[MAX][MAX];  // adj matrix to hold minimum spanning tree of G
+int V;              // number of vertices in V in G
```
Comment (Panquesito7) [inline greedy_approach/prim.c:41]: Avoid using global variables, as they may cause overflows and other issues. Pass them as local function parameters or as struct members.

File: greedy_approach/prim.c [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+ *
+ *
+ * The following link provides a visual representation of graphs that can be used to test / verify the algorithm for different adj
+ * matrices and their weighted, connected graphs.
+ * (https://visualgo.net/en/mst)
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
```
Comment (Panquesito7) [inline greedy_approach/prim.c:34]: Please add a brief description of what the library/header is for (see the example below).
```c++
#include <iostream>   /// for IO operations
#include <cassert>   /// for assert
```

File: greedy_approach/prim.c [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        }
+    }
+}
+
+/**
+ * @brief Function test
+ * uses test matrix and solution of test to verify correctness of prim()
+ * @returns 0:success || -1:fail
+ */
+int test()
```
Comment (Panquesito7) [inline greedy_approach/prim.c:120]: Test functions should be `void` types, and they should use the `assert` function (see the [typical structure](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program for more information).

File: greedy_approach/prim.c [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+          }
+      }
+      return 0;
+}
+
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
```
Comment (Panquesito7) [inline greedy_approach/prim.c:194]:
```suggestion
/**
 * @brief Main function
 * @param argc commandline argument count (ignored)
 * @param argv commandline array of arguments (ignored)
 * @returns 0 on exit
 */
```

File: greedy_approach/prim.c [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+int main(int argc, char const *argv[])
+{   
+    if(argc == 2 && strcmp(argv[1],"-test") == 0)
+    {
+      assert(test() == 0);
```
Comment (Panquesito7) [inline greedy_approach/prim.c:199]: `assert` should be used in the test function directly.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work!

File: greedy_approach/prim.c [hunk 8]
```diff
@@ -0,0 +1,209 @@
+/**
+ * @file
+ * @author [Timothy Maloney](https://github.com/sl1mb0)
+ * @brief [Prim's algorithm](https://en.wikipedia.org/wiki/Prim%27s_algorithm) implementation
+ * implementation of prim's algorithm in C to find the MST of a weighted, connected graph.
```
Comment (Panquesito7) [inline greedy_approach/prim.c:5]:
```suggestion
 * in C to find the MST of a weighted, connected graph.
```

File: greedy_approach/prim.c [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+#define MAX 20
+#define INF 999
+
+int G[MAX][MAX];    // weighted, connected graph G
+int MST[MAX][MAX];  // adj matrix to hold minimum spanning tree of G
+int V;              // number of vertices in V in G
+
+/**
+ * @brief Function minimum
```
Comment (Panquesito7) [inline greedy_approach/prim.c:44]: Please add proper documentation here.

File: greedy_approach/prim.c [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+ * Returns index of minimum element in array composed of edges connected to arbitrary vertex
+ * @param arr graph row
+ * @param N number of elements in arr
+ * @returns index of minimum element in arr
+ */
+int minimum(int arr[], int N)
```
Comment (Panquesito7) [inline greedy_approach/prim.c:50]: Consider using `uint64_t`, `uint32_t`, `uint16_t`, or even `uint8_t` for non-negative values. For negative values, use `int64_t` or their appropriate size ([reference](https://en.cppreference.com/w/c/types/integer)). Requires the `inttypes.h` header. Check out other parts of the code, too. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please address **all** the unresolved conversations above. 🙂

File: greedy_approach/prim.c (hunk 9 above)
Comment (Sl1mb0) [reply greedy_approach/prim.c:41]: What needs to be added? I removed the global variables so I'm not sure what you mean.

Comment (Panquesito7) [reply greedy_approach/prim.c:41]: A brief description of the function is what the functions are for, how does it work, etc. but in short terms. For the detailed description, use `@details`. 🙂

File: greedy_approach/prim.c [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+ *
+ * The following link provides a visual representation of graphs that can be used to test / verify the algorithm for different adj
+ * matrices and their weighted, connected graphs.
+ * (https://visualgo.net/en/mst)
+ */
+
+#include <stdio.h>   /// for IO operations
+#include <string.h>  /// for string comparison
+#include <assert.h>  /// for assert()
+#include <inttypes.h>  /// for uint16_t
```
Comment (Panquesito7) [inline greedy_approach/prim.c:35]:
```suggestion
#include <stdio.h>        /// for IO operations
#include <string.h>      /// for string comparison
#include <assert.h>     /// for assert()
#include <inttypes.h>  /// for uint16_t
```

File: greedy_approach/prim.c [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main(int argc, char const *argv[])
+{   
+
+    uint16_t G[MAX][MAX];    // weighted, connected graph G
+    uint16_t MST[MAX][MAX];  // adj matrix to hold minimum spanning tree of G
+    uint16_t V;              // number of vertices in V in G
```
Comment (Panquesito7) [inline greedy_approach/prim.c:195]:
```suggestion
    uint16_t G[MAX][MAX];    ///< weighted, connected graph G
    uint16_t MST[MAX][MAX];  ///< adj matrix to hold minimum spanning tree of G
    uint16_t V;              ///< number of vertices in V in G
```

**Comment by KirwinD** [comment]
Amazing work.  The results are game changing

**Comment by Sl1mb0** [comment]
Does anything else need to be done?

File: greedy_approach/prim.c [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+            }
+        }
+    }
+}
+
+/**
+ * @brief Function test
+ * uses test matrix and solution of test to verify correctness of prim()
+ * @returns 0:succfess || -1:fail
+ */
```
Comment (Panquesito7) [inline greedy_approach/prim.c:116]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
```

File: greedy_approach/prim.c [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+void test(uint16_t G[][MAX], uint16_t MST[][MAX], uint16_t V)
```
Comment (Panquesito7) [inline greedy_approach/prim.c:117]:
```suggestion
static void test(uint16_t G[][MAX], uint16_t MST[][MAX], uint16_t V)
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by Sl1mb0** [comment]
Okay I went through and made all the requested changes. Thanks for your help @Panquesito7!

**Review by Panquesito7 — APPROVED** [review]
Very nice and amazing work, @Sl1mb0! Thank you for your contribution! 👍 🎉

**Review by ayaankhan98 — APPROVED** [review]

**Pull request merged**